### PR TITLE
[shardformer] support pipeline base vit model 

### LINKED ISSUE #4285: [shardformer] support pipeline for vit
support pipeline for vit

### PR DESCRIPTION
**Pull request by FoolPlayer**
## 📌 Checklist before creating the PR

- [x] I have created an issue for this PR for traceability
- [x] The title follows the standard format: `[doc/gemini/tensor/...]: A concise description`
- [x] I have added relevant tags if possible for us to better distinguish different PRs

## 🚨 Issue number

> Link this PR to your issue with words like fixed to automatically close the linked issue upon merge
>
> e.g. `fixed #1234`, `closed #1234`, `resolved #1234`
fixed #4285 

## 📝 What does this PR do?

> Summarize your work here.
> if you have any plots/diagrams/screenshots/tables, please attach them here.

support pipeline for vit model

## 💥 Checklist before requesting a review

- [x] I have linked my PR to an issue ([instruction](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue))
- [x] My issue clearly describes the problem/feature/proposal, with diagrams/charts/table/code if possible
- [x] I have performed a self-review of my code
- [x] I have added thorough tests.
- [x] I have added docstrings for all the functions/methods I implemented

## ⭐️ Do you enjoy contributing to Colossal-AI?

- [x] 🌝 Yes, I do.
- [ ] 🌚 No, I don't.

Tell us more if you don't enjoy contributing to Colossal-AI.

### CODE REVIEW

**Comment by FoolPlayer** [comment]
image

image

File: colossalai/shardformer/policies/vit.py [hunk 1]
```diff
@@ -15,96 +22,209 @@ def config_sanity_check(self):
...
+                ViTForImageClassification:
+                ModulePolicyDescription(sub_module_replacement=[
+                                        SubModuleReplacementDescription(suffix="classifier",
+                                                                            target_module=Linear1D_Col,
+                                                                            kwargs=dict(gather_output=True))
+                                        ])
+            }
+            policy.update(new_item)
+
+        from transformers.models.vit.modeling_vit import ViTForImageClassification
```
Comment (Fridge003) [inline colossalai/shardformer/policies/vit.py:179]: ViTForImageClassification has been imported in line 165

File: colossalai/shardformer/modeling/vit.py [hunk 2]
```diff
@@ -0,0 +1,337 @@
...
+        output_hidden_states = (output_hidden_states
+                                if output_hidden_states is not None else self.config.output_hidden_states)
+        return_dict = return_dict if return_dict is not None else self.config.use_return_dict
+
+        if output_attentions is not None:
+            logging.warning('Non-empty output_attentions is not supported for pipeline models at the moment.')
+            output_attentions = None
+        if output_hidden_states is not None:
+            logging.warning('Non-empty output_hidden_states is not supported for pipeline models at the moment.')
+            output_hidden_states = None
```
Comment (Fridge003) [inline colossalai/shardformer/modeling/vit.py:85]: use_cache and past_key_values should also be set to False here, please refer to code of gpt2/bert

File: colossalai/shardformer/modeling/vit.py [hunk 3]
```diff
@@ -0,0 +1,337 @@
...
+                create_custom_forward(layer_module),
+                hidden_states,
+                layer_head_mask,
+            )
+        else:
+            layer_outputs = layer_module(hidden_states, layer_head_mask, False)
+
+        hidden_states = layer_outputs[0]
+    if not stage_manager.is_last_stage():
+        return hidden_states
```
Comment (Fridge003) [inline colossalai/shardformer/modeling/vit.py:44]: hidden_states should be a dict

Comment (FoolPlayer) [reply colossalai/shardformer/modeling/vit.py:44]: done

File: colossalai/shardformer/modeling/vit.py (hunk 2 above)
Comment (FoolPlayer) [reply colossalai/shardformer/modeling/vit.py:85]: no cache use and kv-store of vit model

**Review by Fridge003 — APPROVED** [review]

**Pull request merged**